src/NServiceBus.Core/Routing/EndpointInstances.cs

```diff
@@ -3,17 +3,23 @@ namespace NServiceBus.Routing
     using System;
     using System.Collections.Generic;
     using System.Linq;
+    using System.Threading.Tasks;
 
     /// <summary>
     /// Stores the information about instances of known endpoints.
     /// </summary>
     public class EndpointInstances
     {
-        List<Func<Endpoint, IEnumerable<EndpointInstance>>> rules = new List<Func<Endpoint, IEnumerable<EndpointInstance>>>();
+        List<Func<Endpoint, Task<IEnumerable<EndpointInstance>>>> rules = new List<Func<Endpoint, Task<IEnumerable<EndpointInstance>>>>();
 
-        internal IEnumerable<EndpointInstance> FindInstances(Endpoint endpoint)
+        internal async Task<IEnumerable<EndpointInstance>> FindInstances(Endpoint endpoint)
         {
-            var distinctInstances = rules.SelectMany(r => r(endpoint)).Distinct().ToArray();
+            var instances = new List<EndpointInstance>();
+            foreach (var rule in rules)
+            {
+                instances.AddRange(await rule.Invoke(endpoint));
+            }
+            var distinctInstances = instances.Distinct().ToArray();
```

Comment:
Here we do need it because various sources can provide the same instance and downstream code expects the items to be unique.